From 59871d5ead3677294006390cb1e9d9998ddb5bb2 Mon Sep 17 00:00:00 2001
From: Thomas Perl <m@thp.io>
Date: Thu, 5 Feb 2015 18:36:29 +0100
Subject: [PATCH 52/65] Report distro upgrade size in custom request

If carrying out a distro-upgrade, the required size can be determined
via a custom "nemo::query-size:" prefix to the pattern name, e.g.:

    pkcon upgrade-system nemo::query-size:jolla-configuration-sbj complete

The error message will be formatted like this:

    "DOWNLOAD=<download bytes>;INSTALL=<install bytes>;REMOVE=<remove bytes>"

The error ID is PK_ERROR_ENUM_NO_DISTRO_UPGRADE_DATA.
---
 backends/zypp/pk-backend-zypp.cpp | 31 +++++++++++++++++++++++++++++--
 1 file changed, 29 insertions(+), 2 deletions(-)

diff --git a/backends/zypp/pk-backend-zypp.cpp b/backends/zypp/pk-backend-zypp.cpp
index 552baffa5..9a5049602 100644
--- a/backends/zypp/pk-backend-zypp.cpp
+++ b/backends/zypp/pk-backend-zypp.cpp
@@ -1774,7 +1774,7 @@ zypp_perform_execution (PkBackendJob *job, ZYpp::Ptr zypp, PerformType type, gbo
 		}
 
 		ResPool pool = ResPool::instance ();
-		if (pk_bitfield_contain (transaction_flags, PK_TRANSACTION_FLAG_ENUM_SIMULATE)) {
+		if (type != UPGRADE && pk_bitfield_contain (transaction_flags, PK_TRANSACTION_FLAG_ENUM_SIMULATE)) {
 			ret = TRUE;
 
 			//MIL << "simulating" << endl;
@@ -1872,6 +1872,8 @@ zypp_perform_execution (PkBackendJob *job, ZYpp::Ptr zypp, PerformType type, gbo
 
 		PK_ZYPP_LOG("Before commit: %d downloads, %d installs, %d removals",
 				priv->exec.total_downloads, priv->exec.total_installs, priv->exec.total_removals);
+		PK_ZYPP_LOG("Byte sizes: %ld download, %ld install, %ld remove",
+				total_download_bytes, total_install_bytes, total_remove_bytes);
 
 		int64_t required_space_bytes = (total_download_bytes + total_install_bytes - total_remove_bytes);
 		// XXX: This assumes package downloads also end up in rootfs, and that
@@ -1888,6 +1890,21 @@ zypp_perform_execution (PkBackendJob *job, ZYpp::Ptr zypp, PerformType type, gbo
 			goto exit;
 		}
 
+		if (pk_bitfield_contain (transaction_flags, PK_TRANSACTION_FLAG_ENUM_SIMULATE)) {
+			gchar *msg = g_strdup_printf("DOWNLOAD=%ld;INSTALL=%ld;REMOVE=%ld",
+					total_download_bytes, total_install_bytes, total_remove_bytes);
+			PK_ZYPP_LOG("Reporting upgrade size: '%s'", msg);
+			pk_backend_job_error_code (job, PK_ERROR_ENUM_NO_DISTRO_UPGRADE_DATA, "%s\n", msg);
+			g_free(msg);
+
+			PK_ZYPP_LOG("Simulate requested, resetting status");
+			for (ResPool::const_iterator it = pool.begin (); it != pool.end (); ++it) {
+				it->statusReset ();
+			}
+
+			goto exit;
+		}
+
 		ZYppCommitResult result = zypp->commit (policy);
 
 		bool worked = result.allDone();
@@ -4112,7 +4129,16 @@ backend_upgrade_system_thread (PkBackendJob *job, GVariant *params, gpointer use
 {
 	DistUpgrade *parameters = static_cast<DistUpgrade *>(user_data);
 	PkBitfield transaction_flags = 0;
-	std::string pattern_name = std::string("pattern:") + std::string(parameters->distro_id);
+
+	gchar *id = parameters->distro_id;
+
+	if (strstr(id, "nemo::query-size:") == id) {
+		id += strlen("nemo::query-size:");
+		pk_bitfield_add(transaction_flags, PK_TRANSACTION_FLAG_ENUM_SIMULATE);
+		PK_ZYPP_LOG("Getting size of distro upgrade, with pattern = '%s'", id);
+	}
+
+	std::string pattern_name = std::string("pattern:") + std::string(id);
 	bool install_pattern = false;
 
 	/**
@@ -4174,6 +4200,7 @@ backend_upgrade_system_thread (PkBackendJob *job, GVariant *params, gpointer use
 			zypp_get_packages_by_name(pattern_name.c_str(), ResKind::pattern, patterns);
 
 			if (patterns.size() == 0) {
+				PK_ZYPP_LOG ("Pattern not found: %s", pattern_name.c_str());
 				//MIL << "Pattern not found: " << pattern_name << " - ignoring" << std::endl;
 			}
 
-- 
2.13.5

